File: src/_3_medium_longest_substring_without_repeating_characters.ts
import { assertEquals } from "test";
/*
 * @lc app=leetcode id=3 lang=typescript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
function lengthOfLongestSubstring(s: string): number {
  const curr: string[] = [];
  let max = 0;

  for (const c of s) {
    const i = curr.indexOf(c);
    if (i !== -1) {
      for (let j = 0; j <= i; j++) {
        curr.shift();
      }
    }
    curr.push(c);
    max = Math.max(max, curr.length);
  }
  return max;
}
// @lc code=end

Deno.test("lengthOfLongestSubstring", (_) => {
  assertEquals(lengthOfLongestSubstring("pwwkew"), 3);
  assertEquals(lengthOfLongestSubstring("abcabcdbg"), 4);
  assertEquals(lengthOfLongestSubstring("aaaa"), 1);
});

File: src/_3_medium_longest_substring_without_repeating_characters.rs
/*
 * @lc app=leetcode id=3 lang=rust
 *
 * [3] Longest Substring Without Repeating Characters
 */

struct Solution {}

// @lc code=start
use std::cmp::max;
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut curr = Vec::new();
        let mut max_len = 0;

        for c in s.chars() {
            if let Some(i) = curr.iter().position(|&x| x == c) {
                curr.drain(0..i + 1);
            }
            curr.push(c);
            max_len = max(curr.len(), max_len);
        }
        max_len as i32
    }
}
// @lc code=end

#[test]
fn test3() {
    assert_eq!(
        Solution::length_of_longest_substring(String::from("abcabcbb")),
        3
    );
}
